klayout/drc/run_drc.py

```diff
@@ -0,0 +1,709 @@
+            run_check(drc_path, "density", layout_path, drc_run_dir, switches)
+        )
+
+        if arguments["--density_only"]:
+            logging.info("## Completed running density checks only.")
+            exit()
+
+    ## Generate run rule deck from template.
+    if not arguments["--table"]:
+        drc_file = generate_drc_run_template(rule_deck_full_path, drc_run_dir)
+    else:
+        drc_file = generate_drc_run_template(
+            rule_deck_full_path, drc_run_dir, arguments["--table"]
+        )
+
+    ## Run Main DRC
+    list_res_db_files.append(
+        run_check(drc_file, "main", layout_path, drc_run_dir, switches)
+    )
+
+    ## Check run
+    check_drc_results(list_res_db_files)
+
+
+def main(drc_run_dir: str, now_str: str, arguments: dict):
+    """
+    main function to run the DRC.
+
+    Parameters
+    ----------
+    drc_run_dir : str
+        String with absolute path of the full run dir.
+    now_str : str
+        String with the run name for logs.
+    arguments : dict
+        Dictionary that holds the arguments used by user in the run command. This is generated by docopt library.
+    """
+
+    # Check gds file existance
+    if os.path.exists(arguments["--path"]):
+        pass
+    else:
+        logging.error("The input GDS file path doesn't exist, please recheck.")
+        exit()
+
+    rule_deck_full_path = os.path.dirname(os.path.abspath(__file__))
+
+    ## Check Klayout version
+    check_klayout_version()
+
+    ## Check if there was a layout provided.
+    if not arguments["--path"]:
+        logging.error("No provided gds file, please add one")
+        exit(1)
+
+    ## Check layout type
+    layout_path = arguments["--path"]
+    layout_path = check_layout_path(layout_path)
+
+    ## Get run switches
+    switches = generate_klayout_switches(arguments, layout_path)
+
+    if (
+        int(arguments["--mp"]) == 1
+        or arguments["--antenna_only"]
+        or arguments["--density_only"]
+    ):
+        run_single_processor(
+            arguments, rule_deck_full_path, layout_path, switches, drc_run_dir
+        )
+    else:
+        run_parallel_run(
+            arguments, rule_deck_full_path, layout_path, switches, drc_run_dir
+        )
+
+
+# ================================================================
+# -------------------------- MAIN --------------------------------
+# ================================================================
+
+if __name__ == "__main__":
+
+    # arguments
+    arguments = docopt(__doc__, version="RUN DRC: 1.0")
+
+    # logs format
+    now_str = datetime.utcnow().strftime("drc_run_%Y_%m_%d_%H_%M_%S")
+
+    if (
+        arguments["--run_dir"] == "pwd"
+        or arguments["--run_dir"] == ""
+        or arguments["--run_dir"] is None
+    ):
+        drc_run_dir = os.path.join(os.path.abspath(os.getcwd()), now_str)
+    else:
+        drc_run_dir = os.path.abspath(arguments["--run_dir"])
+
+    os.makedirs(drc_run_dir, exist_ok=True)
```

Comment:
should we error out if the directory already exist? (to prevent user to inadvertently overwrite things)

Reply:
No need to do this as we usually generate a new rundir name with a different time stamp with seconds in the time. We only use that switch for manual testing avoiding generating many rundirs

Reply:
maybe then we should just log a warning that things might get overwritten?